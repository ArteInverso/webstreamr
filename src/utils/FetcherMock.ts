/* istanbul ignore file */
import fs from 'node:fs';
import slugify from 'slugify';
import winston from 'winston';
import crypto from 'crypto';
import CachePolicy from 'http-cache-semantics';
import { Context } from '../types';
import { envGet } from './env';
import { Fetcher } from './Fetcher';

export class FetcherMock extends Fetcher {
  public constructor() {
    super(winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] }));
  }

  public override async text(ctx: Context, url: URL, init?: RequestInit): Promise<string> {
    const path = `${__dirname}/__fixtures__/Fetcher/${this.slugifyUrl(url)}`;

    return this.fetch(path, ctx, url, init);
  };

  public override async textPost(ctx: Context, url: URL, body: string, init?: RequestInit): Promise<string> {
    const path = `${__dirname}/__fixtures__/Fetcher/post-${this.slugifyUrl(url)}-${slugify(body)}`;

    return this.fetch(path, ctx, url, { ...init, method: 'POST', body });
  };

  public override async head(ctx: Context, url: URL, init?: RequestInit): Promise<CachePolicy.Headers> {
    const path = `${__dirname}/__fixtures__/Fetcher/head-${this.slugifyUrl(url)}`;

    return JSON.parse(await this.fetch(path, ctx, url, { ...init, method: 'HEAD' }));
  };

  private readonly slugifyUrl = (url: URL): string => {
    if (url.href.length > 255) {
      return slugify(`${url.origin}-${crypto.createHash('md5').update(url.href).digest('hex')}`);
    }

    return slugify(url.href);
  };

  private readonly fetch = async (path: string, ctx: Context, url: URL, init?: RequestInit): Promise<string> => {
    const errorPath = `${path}.error`;

    if (fs.existsSync(errorPath)) {
      throw new Error(fs.readFileSync(errorPath).toString());
    } else if (fs.existsSync(path)) {
      return fs.readFileSync(path).toString();
    } else {
      let response;
      try {
        if (envGet('TEST_UPDATE_FIXTURES')) {
          response = await super.fetchWithTimeout(ctx, url, init);
        } else {
          console.error(`No fixture found at "${path}".`);
          process.exit(1);
        }
      } catch (error) {
        fs.writeFileSync(errorPath, `${error}`);
        throw error;
      }

      if (!response.ok) {
        const message = `Fetcher error: ${response.status}: ${response.statusText}`;
        fs.writeFileSync(errorPath, message);
        throw new Error(message);
      }

      let result;
      if (init?.method === 'HEAD') {
        result = JSON.stringify(response.headers);
      } else {
        result = await response.text();
      }

      fs.writeFileSync(path, result);

      return result;
    }
  };
}
