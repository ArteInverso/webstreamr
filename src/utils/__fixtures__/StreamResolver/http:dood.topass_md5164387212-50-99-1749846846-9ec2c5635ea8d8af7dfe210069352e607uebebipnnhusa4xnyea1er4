https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rlyusvlvslr3u3ta/bibyo6w0dm~