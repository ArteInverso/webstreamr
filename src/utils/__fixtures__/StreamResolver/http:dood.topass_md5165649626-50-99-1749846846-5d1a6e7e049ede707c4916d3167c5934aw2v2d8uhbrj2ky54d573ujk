https://kk892as.cloudatacdn.com/u5kj7j7s27d3sdgge5woezkbi4pu672wktq3aqujw47rbx4declm2mk2k2gq/4msi5kvncy~