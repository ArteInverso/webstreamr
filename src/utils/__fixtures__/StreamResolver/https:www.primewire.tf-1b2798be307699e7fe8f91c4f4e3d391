{
  "link": "https://mixdrop.ag/f/gn1d6nkjiwdp8x4",
  "host_id": 29,
  "host": "mixdrop.ag"
}