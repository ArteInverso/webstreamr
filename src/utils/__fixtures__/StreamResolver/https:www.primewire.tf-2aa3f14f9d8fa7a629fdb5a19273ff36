{
  "link": "https://www.youtube.com/watch?v=Z8ZhQPaw4rE",
  "host_id": 19,
  "host": "youtube.com"
}