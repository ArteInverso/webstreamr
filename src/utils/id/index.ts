import { ImdbId } from './ImdbId';
import { TmdbId } from './TmdbId';
import { getImdbIdFromTmdbId, getTmdbIdFromImdbId } from '../tmdb';
import { Fetcher } from '../Fetcher';
import { Context } from '../../types';

export * from './ImdbId';
export * from './TmdbId';

export type Id = ImdbId | TmdbId;

export const getImdbId = async (ctx: Context, fetcher: Fetcher, id: Id): Promise<ImdbId> => {
  if (id instanceof TmdbId) {
    return await getImdbIdFromTmdbId(ctx, fetcher, id);
  }

  return id;
};

export const getTmdbId = async (ctx: Context, fetcher: Fetcher, id: Id): Promise<TmdbId> => {
  if (id instanceof ImdbId) {
    return await getTmdbIdFromImdbId(ctx, fetcher, id);
  }

  return id;
};
