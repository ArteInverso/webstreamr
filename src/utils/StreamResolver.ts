import { ContentType, Stream } from 'stremio-addon-sdk';
import winston from 'winston';
import bytes from 'bytes';
import { Context, UrlResult } from '../types';
import { Source } from '../source';
import {
  BlockedError,
  HttpError,
  NotFoundError,
  QueueIsFullError,
  TimeoutError,
  TooManyTimeoutsError,
  TooManyRequestsError,
} from '../error';
import { flagFromCountryCode } from './language';
import { envGetAppName } from './env';
import { Id } from './id';
import { ExtractorRegistry } from '../extractor';
import { showExternalUrls } from './config';

interface ResolveResponse {
  streams: Stream[];
  ttl?: number;
}

export class StreamResolver {
  private readonly logger: winston.Logger;
  private readonly extractorRegistry: ExtractorRegistry;

  public constructor(logger: winston.Logger, extractorRegistry: ExtractorRegistry) {
    this.logger = logger;
    this.extractorRegistry = extractorRegistry;
  }

  public async resolve(ctx: Context, sources: Source[], type: ContentType, id: Id): Promise<ResolveResponse> {
    if (sources.length === 0) {
      return {
        streams: [
          {
            name: 'WebStreamr',
            title: '‚ö†Ô∏è No sources found. Please re-configure the plugin.',
            ytId: 'E4WlUXrJgy4',
          },
        ],
      };
    }

    const streams: Stream[] = [];

    let handlerErrorOccurred = false;
    const urlResults: UrlResult[] = [];
    const handlerPromises = sources.map(async (handler) => {
      if (!handler.contentTypes.includes(type)) {
        return;
      }

      try {
        const handleResults = await handler.handle(ctx, type, id);
        this.logger.info(`${handler.id} returned ${handleResults.length} urls`, ctx);

        const handlerUrlResults = await Promise.all(
          handleResults.map(({ countryCode, title, url }) => this.extractorRegistry.handle(ctx, url, countryCode, title)),
        );

        urlResults.push(...handlerUrlResults.flat());
      } catch (error) {
        if (error instanceof NotFoundError) {
          return;
        }

        handlerErrorOccurred = true;

        streams.push({
          name: envGetAppName(),
          title: [`üîó ${handler.label}`, this.logErrorAndReturnNiceString(ctx, handler.id, error)].join('\n'),
          ytId: 'E4WlUXrJgy4',
        });
      }
    });
    await Promise.all(handlerPromises);

    urlResults.sort((a, b) => {
      const heightComparison = (b.meta.height ?? 0) - (a.meta.height ?? 0);
      if (heightComparison !== 0) {
        return heightComparison;
      }

      const bytesComparison = (b.meta.bytes ?? 0) - (a.meta.bytes ?? 0);
      if (bytesComparison !== 0) {
        return bytesComparison;
      }

      if (a.isExternal || b.isExternal) {
        return a.isExternal ? 1 : -1;
      }

      return a.label.localeCompare(b.label);
    });

    this.logger.info(`Return ${urlResults.length} streams`, ctx);

    streams.push(
      ...urlResults.filter(urlResult => !urlResult.isExternal || showExternalUrls(ctx.config) || urlResult.error)
        .map(urlResult => ({
          ...this.buildUrl(ctx, urlResult),
          name: this.buildName(ctx, urlResult),
          title: this.buildTitle(ctx, urlResult),
          behaviorHints: {
            ...(urlResult.sourceId && { bingeGroup: `webstreamr-${urlResult.sourceId}` }),
            ...(urlResult.requestHeaders !== undefined && {
              notWebReady: true,
              proxyHeaders: { request: urlResult.requestHeaders },
            }),
            ...(urlResult.meta.bytes && { videoSize: urlResult.meta.bytes }),
          },
        })),
    );

    const ttl = !handlerErrorOccurred ? this.determineTtl(urlResults) : undefined;

    return {
      streams,
      ...(ttl && { ttl }),
    };
  };

  private determineTtl(urlResults: UrlResult[]): number | undefined {
    if (!urlResults.length) {
      return 900000; // 15m
    }

    if (urlResults.some(urlResult => urlResult.ttl === undefined)) {
      return undefined;
    }

    return Math.min(...urlResults.map(urlResult => urlResult.ttl as number));
  };

  private buildUrl(ctx: Context, urlResult: UrlResult): { externalUrl: string } | { url: string } | { ytId: string } {
    if (!urlResult.isExternal) {
      return { url: urlResult.url.href };
    }

    if (showExternalUrls(ctx.config)) {
      return { externalUrl: urlResult.url.href };
    }

    return { ytId: 'E4WlUXrJgy4' };
  };

  private buildName(ctx: Context, urlResult: UrlResult): string {
    let name = envGetAppName();

    name += ` ${flagFromCountryCode(urlResult.meta.countryCode)}`;

    if (urlResult.meta.height) {
      name += ` ${urlResult.meta.height}p`;
    }

    if (urlResult.isExternal && showExternalUrls(ctx.config)) {
      name += ` ‚ö†Ô∏è external`;
    }

    return name;
  };

  private logErrorAndReturnNiceString(ctx: Context, source: string, error: unknown): string {
    if (error instanceof BlockedError) {
      if (error.reason === 'cloudflare_challenge') {
        this.logger.warn(`${source}: Request was blocked via Cloudflare challenge.`, ctx);
      } else {
        this.logger.warn(`${source}: Request was blocked, headers: ${JSON.stringify(error.headers)}.`, ctx);
      }

      return '‚ö†Ô∏è Request was blocked.';
    }

    if (error instanceof TooManyRequestsError) {
      this.logger.warn(`${source}: Rate limited for ${error.retryAfter} seconds.`, ctx);

      return 'üö¶ Request was rate-limited. Please try again later or consider self-hosting.';
    }

    if (error instanceof TooManyTimeoutsError) {
      this.logger.warn(`${source}: Too many timeouts.`, ctx);

      return 'üö¶ Too many recent timeouts. Please try again later.';
    }

    if (error instanceof TimeoutError) {
      this.logger.warn(`${source}: Request timed out.`, ctx);

      return 'üê¢ Request timed out.';
    }

    if (error instanceof QueueIsFullError) {
      this.logger.warn(`${source}: Request queue is full.`, ctx);

      return '‚è≥ Request queue is full. Please try again later or consider self-hosting.';
    }

    if (error instanceof HttpError) {
      this.logger.error(`${source}: HTTP status ${error.status} (${error.statusText}), headers: ${JSON.stringify(error.headers)}, stack: ${error.stack}.`, ctx);
      return `‚ùå Request failed with status ${error.status} (${error.statusText}). Request-id: ${ctx.id}.`;
    }

    const cause = (error as Error & { cause?: unknown }).cause;
    this.logger.error(`${source} error: ${error}, cause: ${cause}, stack: ${(error as Error).stack}`, ctx);

    return `‚ùå Request failed. Request-id: ${ctx.id}.`;
  };

  private buildTitle(ctx: Context, urlResult: UrlResult): string {
    const titleLines = [];

    if (urlResult.meta.title) {
      titleLines.push(urlResult.meta.title);
    }

    const titleDetailsLine = [];
    if (urlResult.meta.bytes) {
      titleDetailsLine.push(`üíæ ${bytes.format(urlResult.meta.bytes, { unitSeparator: ' ' })}`);
    }
    titleDetailsLine.push(`üîó ${urlResult.label}`);
    titleLines.push(titleDetailsLine.join(' '));

    if (urlResult.error) {
      titleLines.push(this.logErrorAndReturnNiceString(ctx, urlResult.sourceId, urlResult.error));
    }

    return titleLines.join('\n');
  };
}
