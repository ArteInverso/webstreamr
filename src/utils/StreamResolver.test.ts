import { ContentType } from 'stremio-addon-sdk';
import winston from 'winston';
import { createExtractors, Extractor, ExtractorRegistry } from '../extractor';
import { StreamResolver } from './StreamResolver';
import { MeineCloud, MostraGuarda, Source, SourceResult } from '../source';
import { BlockedReason, CountryCode, Format, UrlResult } from '../types';
import { BlockedError, HttpError, NotFoundError, QueueIsFullError, TimeoutError, TooManyRequestsError, TooManyTimeoutsError } from '../error';
import { ImdbId } from './id';
import { FetcherMock } from './FetcherMock';
import { createTestContext } from '../test';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const fetcher = new FetcherMock(`${__dirname}/__fixtures__/StreamResolver`);
const ctx = createTestContext({ de: 'on', it: 'on' });

const meineCloud = new MeineCloud(fetcher);
const mostraGuarda = new MostraGuarda(fetcher);

describe('resolve', () => {
  test('returns info as stream if no sources were configured', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [], 'movie', new ImdbId('tt123456789', undefined, undefined));

    expect(streams).toMatchSnapshot();
  });

  test('returns source errors as stream', async () => {
    const fetcherSpy = jest.spyOn(fetcher, 'text').mockRejectedValue('ups, an error occurred.');
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud], 'movie', new ImdbId('tt123456789', undefined, undefined));

    expect(streams).toMatchSnapshot();

    fetcherSpy.mockRestore();
  });

  test('returns empty array if no source found anything', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'movie', new ImdbId('tt12345678', undefined, undefined));

    expect(streams).toMatchSnapshot();
  });

  test('returns empty array if no source supported the type', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'series', new ImdbId('tt12345678', 1, 1));

    expect(streams).toMatchSnapshot();
  });

  test('returns sorted results', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streams.ttl).not.toBeUndefined();
    expect(streams.streams).toMatchSnapshot();

    const streamsWithExternalUrls = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, includeExternalUrls: 'on' } }, [meineCloud, mostraGuarda], 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streamsWithExternalUrls.ttl).not.toBeUndefined();
    expect(streamsWithExternalUrls.streams).toMatchSnapshot();
  });

  test('adds error info', async () => {
    class MockHandler implements Source {
      public readonly id = 'mockhandler';

      public readonly label = 'MockHandler';

      public readonly contentTypes: ContentType[] = ['movie'];

      public readonly countryCodes: CountryCode[] = [CountryCode.de];

      public readonly handle = async (): Promise<SourceResult[]> => {
        return [{ countryCode: CountryCode.de, url: new URL('https://example.com') }];
      };
    }

    class MockExtractor extends Extractor {
      public readonly id = 'mockextractor';

      public readonly label = 'MockExtractor';

      public override readonly ttl = 1;

      public readonly supports = (): boolean => true;

      protected readonly extractInternal = async (): Promise<UrlResult[]> =>
        [
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.cloudflare_challenge, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.media_flow_proxy_auth, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.unknown, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TooManyRequestsError(10),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TooManyTimeoutsError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example2.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TypeError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example2.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TimeoutError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example3.com'),
            format: Format.unknown,
            isExternal: true,
            error: new QueueIsFullError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
          {
            url: new URL('https://example4.com'),
            format: Format.unknown,
            isExternal: true,
            error: new HttpError(500, 'Internal Server Error', { 'x-foo': 'bar' }),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCode: CountryCode.de,
            },
          },
        ];
    }

    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, [new MockExtractor()]));

    const streams = await streamResolver.resolve(ctx, [new MockHandler()], 'movie', new ImdbId('tt11655566', undefined, undefined));
    expect(streams).toMatchSnapshot();

    const streamsWithExternalUrls = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, includeExternalUrls: 'on' } }, [new MockHandler()], 'movie', new ImdbId('tt11655566', undefined, undefined));
    expect(streamsWithExternalUrls).toMatchSnapshot();
  });

  test('ignores not found errors', async () => {
    const mockHandler: Source = {
      id: 'mockhandler',
      label: 'MockHandler',
      contentTypes: ['movie'],
      countryCodes: [CountryCode.de],
      handle: jest.fn().mockRejectedValue(new NotFoundError()),
    };
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [mockHandler], 'movie', new ImdbId('tt12345678', undefined, undefined));

    expect(streams).toMatchSnapshot();
  });
});
