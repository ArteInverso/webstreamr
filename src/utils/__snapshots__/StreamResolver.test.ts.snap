// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`resolve adds error info 1`] = `
{
  "streams": [
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⚠️ Request was blocked.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⚠️ Request was blocked.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
🚦 Request was rate-limited. Please try again later or consider self-hosting.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
🚦 Too many recent timeouts. Please try again later.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
❌ Request failed. Request-id: id.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
🐢 Request timed out.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⏳ Request queue is full. Please try again later or consider self-hosting.",
      "ytId": "E4WlUXrJgy4",
    },
    {
      "behaviorHints": {},
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
❌ Request failed with status 500 (Internal Server Error). Request-id: id.",
      "ytId": "E4WlUXrJgy4",
    },
  ],
}
`;

exports[`resolve adds error info 2`] = `
{
  "streams": [
    {
      "behaviorHints": {},
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
⚠️ Request was blocked.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
⚠️ Request was blocked.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
🚦 Request was rate-limited. Please try again later or consider self-hosting.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
🚦 Too many recent timeouts. Please try again later.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example2.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
❌ Request failed. Request-id: id.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example2.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
🐢 Request timed out.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example3.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
⏳ Request queue is full. Please try again later or consider self-hosting.",
    },
    {
      "behaviorHints": {},
      "externalUrl": "https://example4.com/",
      "name": "WebStreamr 🇩🇪 ⚠️ external",
      "title": "🔗 hoster.com
❌ Request failed with status 500 (Internal Server Error). Request-id: id.",
    },
  ],
}
`;

exports[`resolve ignores not found errors 1`] = `
{
  "streams": [],
  "ttl": 900000,
}
`;

exports[`resolve returns empty array if no source found anything 1`] = `
{
  "streams": [],
  "ttl": 900000,
}
`;

exports[`resolve returns empty array if no source supported the type 1`] = `
{
  "streams": [],
  "ttl": 900000,
}
`;

exports[`resolve returns info as stream if no sources were configured 1`] = `
{
  "streams": [
    {
      "name": "WebStreamr",
      "title": "⚠️ No sources found. Please re-configure the plugin.",
      "ytId": "E4WlUXrJgy4",
    },
  ],
}
`;

exports[`resolve returns sorted results 1`] = `
[
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_de",
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 🇩🇪 1080p",
    "title": "des-teufels-bad-2024
💾 1.3 GB 🔗 Dropload",
    "url": "https://srv34.dropload.io/hls2/01/00197/lyo2h1snpe5c_h/master.m3u8?t=MZSqs3F338ocbM2eTs8X23s585dWcYuilGrNNcidNas&s=1749846846&e=14400&f=987607&srv=srv27&i=0.0&sp=0&ii=50.99.15.93&p1=srv27&p2=srv27",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_it",
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 Dropload",
    "url": "https://srv34.dropload.io/hls2/01/00200/xsr90y2ltdyx_h/master.m3u8?t=rsJMeR8rtWZxFi_3DOdlt98U0rvAMincCe4-awDQzHE&s=1749846846&e=14400&f=1004116&srv=srv24&i=0.0&sp=0&ii=50.99.15.93&p1=srv24&p2=srv24",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_it",
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 SuperVideo",
    "url": "https://hfs311.serversicuro.cc/hls/,dnzpcjj427g4a3gyvbth53rxqgffbu4vol4o7gg7pskpzb6josghbkwhnisq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_de",
      "videoSize": 1073741824,
    },
    "name": "WebStreamr 🇩🇪 720p",
    "title": "des-teufels-bad-2024
💾 1 GB 🔗 SuperVideo",
    "url": "https://hfs309.serversicuro.cc/hls/,dnzpfi3d27g4a3gyvbmh5klwtl65qh654hyjtgupd6p577fze64mz3bsmdyq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_de",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 791702409,
    },
    "name": "WebStreamr 🇩🇪",
    "title": "des-teufels-bad-2024
💾 755.03 MB 🔗 DoodStream",
    "url": "https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rlyusvlvslr3u3ta/bibyo6w0dm~mocked-random-string?token=7uebebipnnhusa4xnyea1er4&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_it",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
    },
    "name": "WebStreamr 🇮🇹",
    "title": "the-devils-bath-2024-sd-sub-ita
🔗 DoodStream",
    "url": "https://kk892as.cloudatacdn.com/u5kj7j7s27d3sdgge5woezkbi4pu672wktq3aqujw47rbx4declm2mk2k2gq/4msi5kvncy~mocked-random-string?token=aw2v2d8uhbrj2ky54d573ujk&expiry=639837296000",
  },
]
`;

exports[`resolve returns sorted results 2`] = `
[
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_de",
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 🇩🇪 1080p",
    "title": "des-teufels-bad-2024
💾 1.3 GB 🔗 Dropload",
    "url": "https://srv34.dropload.io/hls2/01/00197/lyo2h1snpe5c_h/master.m3u8?t=MZSqs3F338ocbM2eTs8X23s585dWcYuilGrNNcidNas&s=1749846846&e=14400&f=987607&srv=srv27&i=0.0&sp=0&ii=50.99.15.93&p1=srv27&p2=srv27",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_it",
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 Dropload",
    "url": "https://srv34.dropload.io/hls2/01/00200/xsr90y2ltdyx_h/master.m3u8?t=rsJMeR8rtWZxFi_3DOdlt98U0rvAMincCe4-awDQzHE&s=1749846846&e=14400&f=1004116&srv=srv24&i=0.0&sp=0&ii=50.99.15.93&p1=srv24&p2=srv24",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_it",
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 SuperVideo",
    "url": "https://hfs311.serversicuro.cc/hls/,dnzpcjj427g4a3gyvbth53rxqgffbu4vol4o7gg7pskpzb6josghbkwhnisq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_de",
      "videoSize": 1073741824,
    },
    "name": "WebStreamr 🇩🇪 720p",
    "title": "des-teufels-bad-2024
💾 1 GB 🔗 SuperVideo",
    "url": "https://hfs309.serversicuro.cc/hls/,dnzpfi3d27g4a3gyvbmh5klwtl65qh654hyjtgupd6p577fze64mz3bsmdyq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_de",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 791702409,
    },
    "name": "WebStreamr 🇩🇪",
    "title": "des-teufels-bad-2024
💾 755.03 MB 🔗 DoodStream",
    "url": "https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rlyusvlvslr3u3ta/bibyo6w0dm~mocked-random-string?token=7uebebipnnhusa4xnyea1er4&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_it",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
    },
    "name": "WebStreamr 🇮🇹",
    "title": "the-devils-bath-2024-sd-sub-ita
🔗 DoodStream",
    "url": "https://kk892as.cloudatacdn.com/u5kj7j7s27d3sdgge5woezkbi4pu672wktq3aqujw47rbx4declm2mk2k2gq/4msi5kvncy~mocked-random-string?token=aw2v2d8uhbrj2ky54d573ujk&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-external_de",
    },
    "externalUrl": "https://mixdrop.ag/e/3nzwveprim63or6",
    "name": "WebStreamr 🇩🇪 ⚠️ external",
    "title": "🔗 mixdrop.ag",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-external_it",
    },
    "externalUrl": "https://mixdrop.ag/e/vk196d6xfzwwo1",
    "name": "WebStreamr 🇮🇹 ⚠️ external",
    "title": "🔗 mixdrop.ag",
  },
]
`;

exports[`resolve returns source errors as stream 1`] = `
{
  "streams": [
    {
      "name": "WebStreamr",
      "title": "🔗 MeineCloud
❌ Request failed. Request-id: id.",
      "ytId": "E4WlUXrJgy4",
    },
  ],
}
`;
